# TP1 — Multiple Linear Regression
Working in pairs is acceptable; however, I strongly recommend submitting the assignment individually. The deadline for this first assignment is February 16, 2026. Please send your response to razan.mhanna@inria.fr
 only.


**Dataset:** Diabetes (regression)

## Learning objectives
- Construct a design matrix and add an intercept
- Fit a multiple linear regression model (OLS)
- Interpret coefficients, p-values, and $R^2$
- Make predictions with confidence intervals
- Perform basic diagnostic checks


## Question 1 — Load the dataset
1. Load the **Diabetes** regression dataset.
2. Convert it to a pandas DataFrame.
3. Identify:
   - the number of observations $n$
   - the number of predictors $p$

👉 *Write the code below.*

In [ ]:
!pip -q install statsmodels scikit-learn pandas matplotlib
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import statsmodels.api as sm
from sklearn.datasets import load_diabetes
pd.set_option('display.precision', 4)
print('Ready!')

Ready!


In [ ]:
diab = load_diabetes(as_frame=True)
df = diab.frame.copy()

In [ ]:
df.shape
df

,age,sex,bmi,bp,s1,s2,s3,s4,s5,s6,target
0,0.0381,0.0507,0.0617,0.0219,-0.0442,-0.0348,-0.0434,-0.0026,0.0199,-0.0176,151.0
1,-0.0019,-0.0446,-0.0515,-0.0263,-0.0084,-0.0192,0.0744,-0.0395,-0.0683,-0.0922,75.0
2,0.0853,0.0507,0.0445,-0.0057,-0.0456,-0.0342,-0.0324,-0.0026,0.0029,-0.0259,141.0
3,-0.0891,-0.0446,-0.0116,-0.0367,0.0122,0.0250,-0.0360,0.0343,0.0227,-0.0094,206.0
4,0.0054,-0.0446,-0.0364,0.0219,0.0039,0.0156,0.0081,-0.0026,-0.0320,-0.0466,135.0
...,...,...,...,...,...,...,...,...,...,...,...
437,0.0417,0.0507,0.0197,0.0597,-0.0057,-0.0026,-0.0287,-0.0026,0.0312,0.0072,178.0
438,-0.0055,0.0507,-0.0159,-0.0676,0.0493,0.0792,-0.0287,0.0343,-0.0181,0.0445,104.0
439,0.0417,0.0507,-0.0159,0.0173,-0.0373,-0.0138,-0.0250,-0.0111,-0.0469,0.0155,132.0
440,-0.0455,-0.0446,0.0391,0.0012,0.0163,0.0153,-0.0287,0.0266,0.0445,-0.0259,220.0


`Answer:` 442 observations and 10 predictors (target is not a prediction)




## Question 2 — Define $X$ and $y$
1. Define the response variable $y$.
2. Define the predictor matrix $X$.
3. Add an intercept column to $X$.

👉 *Explain why adding a constant is necessary when using statsmodels.*

In [ ]:
y = df['target']
X = df.drop(columns="target")
X_design = sm.add_constant(X)

`Answer:` Adding a constant is necessary because, without it, you force the model to pass through the origin (0,0). This is usually unrealistic and leads to biased coefficients because the model 'tilts' the line to hit zero instead of fitting the actual data trend.

## Question 3 — Fit the multiple linear regression model
1. Fit an **OLS** model using statsmodels.
2. Display the full regression summary.

👉 *Write down the estimated regression equation.*

In [ ]:
model = sm.OLS(y,X_design)
results = model.fit()
results.params
results.summary()


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 target   R-squared:                       0.518
Model:                            OLS   Adj. R-squared:                  0.507
Method:                 Least Squares   F-statistic:                     46.27
Date:                Tue, 10 Feb 2026   Prob (F-statistic):           3.83e-62
Time:                        11:14:40   Log-Likelihood:                -2386.0
No. Observations:                 442   AIC:                             4794.
Df Residuals:                     431   BIC:                             4839.
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const        152.1335      2.576     59.061      0.000     147.071     157.196
age          -10.0099     59.749     -0.168      0.867    -127.446     107.426
sex         -239.8156     61.222     -3.917      0.000    -360.147    -119.484
bmi          519.8459     66.533      7.813      0.000     389.076     650.616
bp           324.3846     65.422      4.958      0.000     195.799     452.970
s1          -792.1756    416.680     -1.901      0.058   -1611.153      26.802
s2           476.7390    339.030      1.406      0.160    -189.620    1143.098
s3           101.0433    212.531      0.475      0.635    -316.684     518.770
s4           177.0632    161.476      1.097      0.273    -140.315     494.441
s5           751.2737    171.900      4.370      0.000     413.407    1089.140
s6            67.6267     65.984      1.025      0.306     -62.064     197.318
==============================================================================
Omnibus:                        1.506   Durbin-Watson:                   2.029
Prob(Omnibus):                  0.471   Jarque-Bera (JB):                1.404
Skew:                           0.017   Prob(JB):                        0.496
Kurtosis:                       2.726   Cond. No.                         227.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

`Answer`

Based on the OLS regression results, the estimated model is:

$$
\begin{aligned}
\widehat{\text{target}} = &152.13 \\
&- 10.01(\text{age}) \\
&- 239.82(\text{sex}) \\
&+ 519.85(\text{bmi}) \\
&+ 324.38(\text{bp}) \\
&- 792.18(\text{s1}) \\
&+ 476.74(\text{s2}) \\
&+ 101.04(\text{s3}) \\
&+ 177.06(\text{s4}) \\
&+ 751.27(\text{s5}) \\
&+ 67.63(\text{s6})
\end{aligned}
$$

## Question 4 — Interpretation of coefficients
Using the regression output:
1. Which predictors are statistically significant at the 5% level?
2. Interpret **one significant coefficient** in words.
3. Is the intercept meaningful in this context? Explain.

`Answer 1:` const, sex, bmi, bp, s5<br/>
`Answer 2:` bmi (body mass index) is statistically significant and is estimated as having a value of B = 519.8459 with a standard error of 66.533. The value of B is with 95% confindece inside the interval of 389.076 	650.616<br/>
`Answer 3:` Yes the intercept is meaningful, it ensures the model a optimal linear fit to the data. This we can see by the fact that it is statistically significant. However it has no clear interpretation.

## Question 5 — Goodness of fit
1. What is the value of $R^2$?<br/>
`Answer:`  0.518<br/>
2. What does $R^2$ represent in this model?<br/>
`Answer:`The R² represents how much of the variability<br/>  
3. Compare $R^2$ and adjusted $R^2$. Why are they different?<br/>
`Answer:`The Adjusted R² is a modified version of R² where it has been adjusted for the number of independent variables.

## Question 6 — Multicollinearity
1. Compute the correlation matrix of the predictors.
2. Identify any strongly correlated variables.

👉 *Explain how multicollinearity affects coefficient estimates.*

In [ ]:
#1 Computing the correlation matrix of the predictors
corr_matrix = X.corr()

In [ ]:
for col in corr_matrix:
  for row in corr_matrix:
    if (corr_matrix[col][row] >= 0.5) and col > row:
      print(col, "and", row, corr_matrix[col][row])

s2 and s1 0.8966629578104894
s4 and s1 0.5422072805232409
s4 and s2 0.6598168886666438
s5 and s1 0.5155029243689457
s5 and s4 0.6178589739937282


`Answer:`The strong correlation between s1 and s2 (r ≈ 0.90) indicates significant multicollinearity, which destabilizes coefficient estimates by inflating their standard errors and making them highly sensitive to small changes in the data. This redundancy prevents the model from isolating the unique effect of each predictor, often resulting in statistically insignificant p-values or logically incorrect signs for the coefficients. Essentially, when predictors are this closely linked, the regression cannot accurately determine which specific variable is driving the response.

## Question 7 (Bonus) — Global significance of the model
1. State the null hypothesis tested by the **F-statistic**.
2. Based on the p-value, is the model globally significant?
3. What does this tell you about the predictors as a group?

In [ ]:
# YOUR CODE HERE


In [ ]:
# YOUR CODE HERE
